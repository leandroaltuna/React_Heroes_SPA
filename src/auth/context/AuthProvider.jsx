import { useReducer } from 'react';
import { AuthContext } from './AuthContext';
import { authReducer } from './authReducer';
import { types } from '../types';


//* El init es el tercer param donde se puede inicializar el state.
const init = () => {
    
    const userStoraged = JSON.parse( localStorage.getItem( 'userStoraged' ) );

    return {
        logged: !!userStoraged, //? Doble negacion para saber si hay info del usuario almacenado en el localstorage.
        userInfo: userStoraged,
    }

}

// const initialState = {
//     logged: false,
// }

export const AuthProvider = ({ children }) => {
  
    //* El init es el tercer param del useReducer hook que puede ser opcional. En este caso lo usamos para poder guardar los datos del usuario en el localstorage.
    const [ authState, dispatch ] = useReducer( authReducer, {}, init );
    // const [ authState, dispatch ] = useReducer( authReducer, initialState );

    const login = ( name = '' ) => {
        
        const userObj = { id: 'ABC', name: name, };

        const action = {
            type: types.login,
            payload: userObj,
            // payload: {
            //     id: 'ABC',
            //     name: name,
            // }
        };

        localStorage.setItem( 'userStoraged', JSON.stringify( userObj ) );

        dispatch( action );

    }

    const logout = () => {
        
        localStorage.removeItem( 'userStoraged' );

        const action = { type: types.logout, };

        dispatch( action );

    }

    return (
        <AuthContext.Provider 
            value={
                {
                    ...authState,
                    login,
                    logout,
                }
            }
        >
            { children }
        </AuthContext.Provider>
    )

}
