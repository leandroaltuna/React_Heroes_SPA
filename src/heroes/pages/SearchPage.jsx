import { useLocation, useNavigate } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks';
import { getHeroesByName } from '../helpers';
import { HeroItem } from '../components/HeroItem';


export const SearchPage = () => {
    
    //* Obtener la navegacion
    const navigate = useNavigate();
    //* Obtener informacion de donde nos encontramos
    const location = useLocation();

    //* Cambio el nombre de q a queryParam y lo asigno '' por defecto. query-string es una libreria para obtener los parametros que se envian por url.
    const { q : queryParam = '' } = queryString.parse( location.search );
    const heroes = getHeroesByName( queryParam );

    //* Las condiciones ya regresan un valor booleano, por ende no es necesario asignar el valor booleano en el codigo.
    const showSearchAlert = ( queryParam.length === 0 );
    const showErrorAlert = ( queryParam.length > 0 ) && heroes.length === 0;


    const { searchText, onInputChange } = useForm({
        // searchText: '',
        searchText: queryParam,
    });
    

    const onSearchSubmit = ( event ) => {
        
        event.preventDefault();

        // if ( searchText.trim().length <= 1 ) return;

        navigate( `?q=${ searchText }` );

    }
    
    
    return (
        <>
            <h1>Search</h1>
            <hr />

            <div className="row">

                <div className="col-5">
                    
                    <h4>Searching</h4>
                    <hr />

                    <form onSubmit={ onSearchSubmit }>
                        <input
                            className="form-control" 
                            type="text" 
                            name="searchText" 
                            placeholder="Search a hero"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ onInputChange }
                        />

                        <button className="btn btn-outline-primary mt-2">
                            Search
                        </button>
                    </form>

                </div>

                <div className="col-7">
                    
                    <h4>Results</h4>
                    <hr />

                    
                    {/* //* Si no busca nada se mostrara el search alert, caso contrario se oculta. */}
                    <div 
                        className="alert alert-primary animate__animated animate__fadeIn" 
                        style={{ display: (showSearchAlert) ? '' : 'none' }}
                    >
                        Search a hero
                    </div>
                   
                    {/* //* Si la busqueda no retorna ningun valor se mostrara el error alert, caso contrario se oculta. */}
                    <div 
                        className="alert alert-danger animate__animated animate__fadeIn" 
                        style={{ display: (showErrorAlert) ? '' : 'none' }}
                    >
                        No Hero with <b>{ queryParam }</b>
                    </div>

                    {
                        heroes.map( ( hero ) => (
                            <HeroItem 
                                key={ hero.id }
                                { ...hero }
                            />
                        ))
                    }

                </div>

            </div>
        </>
    )
}
