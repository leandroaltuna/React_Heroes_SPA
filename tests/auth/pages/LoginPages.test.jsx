import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { AuthContext } from '../../../src/auth/context';
import { LoginPage } from '../../../src/auth/pages/LoginPage';


const mockedUseNavigate = jest.fn();

jest.mock( 'react-router-dom', () => ({

    ...jest.requireActual( 'react-router-dom' ),
    useNavigate: () => mockedUseNavigate,

}));

describe( 'Testing <LoginPage />', () => { 
    
    beforeEach( () => jest.clearAllMocks() );

    test( 'Should show default values', () => { 
        
        const contextValue = { logged: false };

        const { container } = render(
            <MemoryRouter>
                <AuthContext.Provider value={ contextValue }>
                    <LoginPage />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        expect( container ).toMatchSnapshot();
        
    });

    test( 'Should call onLogin method with login, navigate and localstorage functions', () => { 
        
        const contextValue = { 
            logged: false,
            login: jest.fn(),
        };

        Storage.prototype.getItem = jest.fn();

        render(
            <MemoryRouter>
                <AuthContext.Provider value={ contextValue }>
                    <LoginPage />
                </AuthContext.Provider>
            </MemoryRouter>
        );

        // screen.debug();

        const btnLogin = screen.getByRole( 'button' );
        fireEvent.click( btnLogin );

        expect( contextValue.login ).toHaveBeenCalled();
        expect( contextValue.login ).toHaveBeenCalledWith( 'Chezna' );
        
        expect( mockedUseNavigate ).toHaveBeenCalledWith( '/', { "replace": true } );
        
        expect( localStorage.getItem ).toHaveBeenCalledWith( 'lastPath' );
        
    });

    
});