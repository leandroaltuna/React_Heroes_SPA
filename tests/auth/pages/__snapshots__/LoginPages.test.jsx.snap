// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing <LoginPage /> Should show default values 1`] = `
<div>
  <div
    class="container mt-5"
  >
    <h1>
      LoginPage
    </h1>
    <hr />
    <button
      class="btn btn-primary"
    >
      Login
    </button>
  </div>
</div>
`;
